lib/pure/asyncdispatch.nim

```diff
@@ -1851,6 +1909,105 @@ proc send*(socket: AsyncFD, data: string,
 
   return retFuture
 
+proc close(vd: VirtualEventDispatcher) =
+  assert vd.virtualHandles.len == 0, "Cannot de-Init Virtual Event Dispacter. There are still Pending Events."
+  vd.virtualMuxHandle.unregister()
+  vd.virtualMuxHandle.close()
+  vd.virtualMuxHandle = nil
+  vd.nextVirtualHandle = 0.VirtualFD
+
+proc unregister*(ev: VirtualAsyncEvent) =
+  ## Unregisters event ``ev``.
+  doAssert(ev.vFD != InvalidVirtualFD, "Event is not registered in the queue!")
+
+  let vd = ev.vd
+  withLockIfThreads ev.eventLock:
+    ev.vd.virtualHandles.del ev.vFD
+    ev.vd = nil
+    ev.cb = nil
+    ev.vFD = InvalidVirtualFD
+
+  if vd.virtualHandles.len == 0:
+    # lazy close the physical event with the Dispatcher
+    # The main reason we need this is to make
+    # hasPendingOperations still work
+    # without modifying the ioselector code.
+    vd.close
+
+proc initVirtualEventDispatcher(p: PDispatcher) =
+  p.vd.virtualMuxHandle = newAsyncEvent()
+
+  proc virtualMuxCB(fd: AsyncFD): bool {.closure,gcsafe.} =
+    # find the virtual events that triggered the physical event and
+    # add them to the callback list.
+    # Not very efficient, but requires the least coordination between threads.
+    
+    proc vcbFactory(vfd: VirtualFD) : proc() {.closure, gcsafe.} =
+      let ev = p.vd.virtualHandles[vfd]
+      result = proc() {.closure,gcSafe.} =
+        if ev.cb(vfd.AsyncFD):
+          # the convention is that if the callback returns true, we unregister the event.
+          ev.unregister
+
+    for vfd, ev in p.vd.virtualHandles:
+      withLockIfThreads ev.eventLock:
+        if ev.triggered:
+          ev.triggered = false
+          p.callbacks.addLast(vcbFactory(vfd))
+
+    # always return false b/c we never want to unregister this event
+    return false
+
+  p.vd.virtualMuxHandle.addEvent(virtualMuxCB)
+
+proc newVirtualAsyncEvent*(): VirtualAsyncEvent =
+  ## Creates a new thread-safe ``VirtualAsyncEvent`` object.
+  ##
+  ## New ``VirtualAsyncEvent`` object is not automatically registered with
+  ## dispatcher like ``AsyncSocket``.
+  result = cast[VirtualAsyncEvent](allocShared0(sizeof(VirtualAsyncEventImpl)))
+  result.cb = nil
+  result.vFD = InvalidVirtualFD
+  when compileOption("threads"):
+    initRLock result.eventLock
+
+proc trigger*(ev: VirtualAsyncEvent) =
+  ## Sets new ``VirtualAsyncEvent`` to signaled state.
+  withLockIfThreads ev.eventLock:
+    if ev.vFD == InvalidVirtualFD:
+      # triggering on an event that is not registered is a noop.
+      return
+    ev.triggered = true
+
+    # send the signal to wake up the dispatcher thread.
+    trigger(ev.vd.virtualMuxHandle)
+
+proc addEvent*(ev: VirtualAsyncEvent, cb: Callback) =
```

Comment:
Argh, I know you're being consistent with the AsyncEvent API here, but it annoys me that this API isn't consistent with the socket API...

In any case, it might be worth to add a ``signal`` async proc that works for both ``VirtualAsyncEvent`` and ``AsyncEvent``. We aim to avoid callbacks after all...

Reply:
This is arguably a lower level api that you wouldn't use in async level code.
The FlowVar -> Promise integration for example. A user never actually sees `AsyncEvent`.

IDK if it's necessary, but it might be a nice convenience to have some api that wraps a `AsyncEvent` / `VirtualAsyncEvent` in a Promise for you.

I'd rather focus on getting the fundamentals in and making that a separate PR, since I consider that sugar.

Reply:
It's important to think about how the API will evolve, remember that once this is released we won't be able to change it in a breaking manner.